File: src/main/java/util/Encrypt.java
package util;

import model.Category;
import model.Difficulty;
import model.Question;

import java.util.List;

public class Encrypt {

    // Encrypts a string using a Caesar cipher
    public static String encrypt(String input, int shift) {
        StringBuilder encrypted = new StringBuilder();
        for (int i = 0; i < input.length(); i++) {
            char c = input.charAt(i);
            if (Character.isLetter(c)) {
                char base = Character.isLowerCase(c) ? 'a' : 'A';
                c = (char) (base + (c - base + shift) % 26);
            }
            else if (Character.isDigit(c)) {
                c = (char) ('0' + (c - '0' + shift) % 10);
            }
            encrypted.append(c);
        }
        return encrypted.toString();
    }

    // Decrypts a string using a Caesar cipher
    public static String decrypt(String input, int shift) {
        StringBuilder decrypted = new StringBuilder();
        for (int i = 0; i < input.length(); i++) {
            char c = input.charAt(i);
            if (Character.isLetter(c)) {
                char base = Character.isLowerCase(c) ? 'a' : 'A';
                c = (char) (base + (c - base - shift + 26) % 26);
            }
            else if (Character.isDigit(c)) {
                c = (char) ('0' + (c - '0' - shift + 10) % 10);
            }
            decrypted.append(c);
        }
        return decrypted.toString();
    }

    // Encrypts a question using a Caesar cipher
    public static Question encryptQuestion(Question question, int shift) {
        Category category = question.getCategory();
        String type = question.getType();
        Difficulty difficulty = question.getDifficulty();
        String encryptedQuestion = encrypt(question.getQuestion(), shift);
        String encryptedCorrectAnswer = encrypt(question.getCorrect_answer(), shift);
        List<String> encryptedIncorrectAnswers = question.getIncorrectAnswers();
        for (int i = 0; i < question.getIncorrectAnswers().size(); i++) {
            encryptedIncorrectAnswers.set(i, encrypt(question.getIncorrectAnswers().get(i), shift));
        }
        Question newQuestion = new Question(category, type, difficulty, encryptedQuestion, encryptedCorrectAnswer, encryptedIncorrectAnswers);
        return new Question();
    }

    // Decrypts a question using a Caesar cipher
    public static Question decryptQuestion(Question encryptedQuestion, int shift) {
        String decryptedQuestion = decrypt(encryptedQuestion.getQuestion(), shift);
        String decryptedCorrectAnswer = decrypt(encryptedQuestion.getCorrect_answer(), shift);
        String[] decryptedIncorrectAnswers = new String[encryptedQuestion.getIncorrect_answers().length];
        for (int i = 0; i < encryptedQuestion.getIncorrect_answers().length; i++) {
            decryptedIncorrectAnswers[i] = decrypt(encryptedQuestion.getIncorrect_answers()[i], shift);
        }
        return new Question(encryptedQuestion.getCategory(), encryptedQuestion.getType(),
                encryptedQuestion.getDifficulty(), decryptedQuestion,
                decryptedCorrectAnswer, decryptedIncorrectAnswers);
    }
}

File: src/main/webapp/scripts/encrypt.js
import { Difficulty } from './difficulty.js';
import { Question } from './question.js';

class Encrypt {
    // Encrypts a string using a Caesar cipher
    static encrypt(input, shift) {
        let encrypted = '';
        for (let i = 0; i < input.length; i++) {
            let c = input.charAt(i);
            if (/[a-zA-Z]/.test(c)) {
                let base = c.toLowerCase() === c ? 'a'.charCodeAt(0) : 'A'.charCodeAt(0);
                c = String.fromCharCode(base + ((c.charCodeAt(0) - base + shift) % 26));
            } else if (/[0-9]/.test(c)) {
                c = String.fromCharCode('0'.charCodeAt(0) + ((c.charCodeAt(0) - '0'.charCodeAt(0) + shift) % 10));
            }
            encrypted += c;
        }
        return encrypted;
    }

    // Decrypts a string using a Caesar cipher
    static decrypt(input, shift) {
        let decrypted = '';
        for (let i = 0; i < input.length; i++) {
            let c = input.charAt(i);
            if (/[a-zA-Z]/.test(c)) {
                let base = c.toLowerCase() === c ? 'a'.charCodeAt(0) : 'A'.charCodeAt(0);
                c = String.fromCharCode(base + ((c.charCodeAt(0) - base - shift + 26) % 26));
            } else if (/[0-9]/.test(c)) {
                c = String.fromCharCode('0'.charCodeAt(0) + ((c.charCodeAt(0) - '0'.charCodeAt(0) - shift + 10) % 10));
            }
            decrypted += c;
        }
        return decrypted;
    }

    // Encrypts a question using a Caesar cipher
    static encryptQuestion(question, shift) {
        const { category, type, difficulty } = question;
        const encryptedQuestion = Encrypt.encrypt(question.question, shift);
        const encryptedCorrectAnswer = Encrypt.encrypt(question.correct_answer, shift);
        const encryptedIncorrectAnswers = question.incorrect_answers.map(answer => Encrypt.encrypt(answer, shift));
        return new Question(category, type, Difficulty[difficulty], encryptedQuestion, encryptedCorrectAnswer, encryptedIncorrectAnswers);
    }

    // Decrypts a question using a Caesar cipher
    static decryptQuestion(encryptedQuestion, shift) {
        const decryptedQuestion = Encrypt.decrypt(encryptedQuestion.question, shift);
        const decryptedCorrectAnswer = Encrypt.decrypt(encryptedQuestion.correct_answer, shift);
        const decryptedIncorrectAnswers = encryptedQuestion.incorrect_answers.map(answer => Encrypt.decrypt(answer, shift));
        return new Question(encryptedQuestion.category, encryptedQuestion.type, Difficulty[encryptedQuestion.difficulty], decryptedQuestion, decryptedCorrectAnswer, decryptedIncorrectAnswers);
    }
}

export { Encrypt };
